Урок-04: Консольные утилиты (Vue CLI)

GoncharSY opened the pull request:
- Все четыре задачи решены;
- Авто-тесты проходит;
- Визуально похоже на примеры, что в условиях задач.

Четвертая задача, с календарем, понравилась. Почти весь день над ней сидел, решал. Этим вечером календарь работает - моральное удовлетворение получено, и настроение хорошее.

GoncharSY committed:
Решил первую задачу (группы полей формы)

GoncharSY committed:
Решил вторую задачу (всплывающие сообщения)

GoncharSY committed:
Решил третью задачу (поле со списком опций)

GoncharSY committed:
Решил четвертую задачу (календарь)

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

GoncharSY committed:
Merge branch 'master' into lesson-04-vue-cli

jsru commented:
Добавляю преподавателя (@ShGKme) для код-ревью.

ShGKme commented on 04-vue-cli/02-TheToaster/components/TheToast.vue:
```diff
@@ -0,0 +1,90 @@
+import UiIcon from './UiIcon.vue';
+
+export default {
+  name: 'TheToast',
```
Имена с префиксом `The` обычно используют для компонентов, которые выводятся на странице один раз. Артикул `The` в англ. показывает, что это "конкретный" компонент, что у него будет "один конкретный экземпляр".

GoncharSY replied:
Переименовал компонент в `UiToast`.

ShGKme commented on 04-vue-cli/03-UiDropdown1/components/UiDropdown.vue:
```diff
@@ -25,6 +29,69 @@ export default {
+    },
+  },
+
+  data() {
+    return {
+      opened: false,
+      iconed: false,
+      option: null,
+    };
+  },
+
+  watch: {
+    modelValue: {
+      immediate: true,
+      handler(nv) {
+        if (!nv) this.option = null;
+        else {
+          this.option = this.options.find((opt) => opt.value === nv);
+          if (!this.option) throw new Error('unknown option');
+        }
+      },
+    },
+
+    options: {
+      immediate: true,
+      deep: true,
+      handler() {
+        this.iconed = this.options.some((opt) => !!opt.icon);
+
+        if (!this.option) return;
+        if (!this.options.includes(this.option)) {
+          this.option = null;
+          this.$emit('update:modelValue', '');
+        }
+      },
+    },
+  },
```
`opened` - собственное состояние компонента. Компонент сам определяет, когда он открыт, а когда закрыт.

Но `iconed` и `option` - не собственное состояние. Эти свойства хранят только результат вычисления от входных параметров. `iconed` явно результат вычисления от `options`. `option` - от `options` + `modelValue`. Сейчас это реализовано вручную, с отслеживанием, но лучше это явно описать именно как вычисляемое свойство.

GoncharSY replied:
Переделал,  и количество кода сократилось. Сейчас это стало очевидным, но когда первый раз решал, почему-то сделал через `watch`. Наверное, взгляд замылился просто.

ShGKme commented on 04-vue-cli/03-UiDropdown1/components/UiDropdown.vue:
```diff
@@ -25,6 +29,69 @@ export default {
+    modelValue: {
+      immediate: true,
+      handler(nv) {
+        if (!nv) this.option = null;
```
`modelValue` может быть `falsy`, но это не значит, что никакой вариант не выбран. Значение какого-то варианта вполне может быть, например, пустой строкой.

GoncharSY replied:
Исправил.

ShGKme commented on 04-vue-cli/03-UiDropdown1/components/UiDropdown.vue:
```diff
@@ -25,6 +29,69 @@ export default {
+        if (!nv) this.option = null;
+        else {
+          this.option = this.options.find((opt) => opt.value === nv);
+          if (!this.option) throw new Error('unknown option');
```
Если передали `modelValue`, для которого нет `option` - это лишь означает, что никакой вариант из представленных не выбран. Лучше не выбрасывать в таких случаях исключения.

GoncharSY replied:
Исправил. На момент решения, подумалось, что так было бы правильней.

ShGKme commented on 04-vue-cli/04-MeetupsCalendar/components/MeetupsCalendar.vue:
```diff
@@ -1,304 +1,117 @@
-  }
-}
+    methods: {
+      distributeEvents(cells, events) {
+        cells = cells || this.cells.filter(c => c.active);
+        events = events || this.monthEvents;
+
+        cells.forEach(cell => {
+          cell.events = events.filter(event => {
+            let cD = new Date(cell.date).toLocaleDateString();
+            let eD = new Date(event.date).toLocaleDateString();
+            return cD === eD;
+          });
+        });
+      },
```
Этот метод мутирует результат вычисления вычисляемого свойства. Ни в коем случае нельзя мутировать вычисляемые свойства, они только для чтения. В `cells` всегда должен быть именно результат функции его вычисления. В любой момент времени Vue может перевычислить значение вычисляемого свойства. (И он может это делать, ведь вычисление не должно никак влиять на значение в том же состоянии).

GoncharSY replied:
Переделал алгоритм распределения митапов по дням в календаре. Убрал метод `distributeEvents`, и теперь массив с митапами за месяц передаю в компоненты ячеек (`CalendarCell`). Внутри ячеек фильтрую митапы для одного дня. Пока что ничего лучше не придумал... Если есть более правильный способ, дайте подсказку, пожалуйста.

ShGKme replied:
Можно один раз пробежаться по данным, чтобы собрать более удобную структуру получения данных по дате. Например, собрать объект, где ключ — дата, а значение — массив данных для этой даты. Тогда можно будет сразу выводить данные для конкретного дня, не перебирая снова и снова циклами одни и те же данные для поиска данных текущего дня.

GoncharSY committed:
Переименовал компонент всплывающего сообщения

GoncharSY committed:
Сделал свойство 'iconed' вычисляемым

GoncharSY committed:
Сделал свойство 'option' вычисляемым

Теперь установка в 'modelValue' значения, которого нет в списке 'options' не вызовет ошибку 'unknown option'.

Теперь falsy значения в 'modelValue' являются допустимыми при поиске элемента в списке 'options'.

GoncharSY committed:
Переделал способ распределения митапов по дням

Теперь календарь выделяет события месяца (monthEvents) и этот массив передает в компоненты ячеек (дней). Ячейки, в свою очередь, самостоятельно вычисляют из месячного списка свои массивы дневных событий (dayEvents) и отображают только их.

jsru commented:
Добавляю преподавателя (@ShGKme) для код-ревью.

ShGKme reviewed (approved):
Все задачи приняты 🔥
